net: fix GetListenPort() to derive the proper port

vasild opened the pull request:
`GetListenPort()` uses a simple logic: "if `-port=P` is given, then we
must be listening on `P`, otherwise we must be listening on `8333`".
This is however not true if `-bind=` has been provided with `:port` part
or if `-whitebind=` has been provided. Thus, extend `GetListenPort()` to
return the port from `-bind=` or `-whitebind=`, if any.

Also, if `-bind=` is provided then we would bind only to a particular address
and should not add all the other addresses of the machine to the list of
local addresses.

Fixes https://github.com/bitcoin/bitcoin/issues/20184

vasild committed:
net: fix GetListenPort() to derive the proper port

`GetListenPort()` uses a simple logic: "if `-port=P` is given, then we
must be listening on `P`, otherwise we must be listening on `8333`".
This is however not true if `-bind=` has been provided with `:port` part
or if `-whitebind=` has been provided. Thus, extend `GetListenPort()` to
return the port from `-bind=` or `-whitebind=`, if any.

Fixes https://github.com/bitcoin/bitcoin/issues/20184 (cases 1. 2. 3. 5.)

vasild committed:
net: only assume all local addresses if listening on any

If `-bind=` is provided then we would bind only to a particular address
and should not add all the other addresses of the machine to the list of
local addresses.

Fixes https://github.com/bitcoin/bitcoin/issues/20184 (case 4.)

MarcoFalke commented on test/functional/feature_bind_port_externalip.py:
```diff
@@ -0,0 +1,76 @@
+    def run_test(self):
+        for i in range(len(expected)):
+            expected_port = expected[i][1]
+            found = False
+            for local in self.nodes[i].getnetworkinfo()['localaddresses']:
+                if local['address'] == '2.2.2.2':
+                    assert_equal(local['port'], expected_port)
+                    found = True
+                    break
+            assert(found)
```
I believe this can be written shorter

```suggestion
            local = next(l for l in self.nodes[i].getnetworkinfo()['localaddresses'] if l['address'] == '2.2.2.2')
            assert_equal(local['port'], expected_port)
```

vasild replied:
To me that is less readable. Maybe my python level is poor, but then maybe other readers of this are also poor pythoners. I prefer to keep it simple stupid:
```
for ...
    if ...
        break
```
that is nearly the same in any programming language.

MarcoFalke commented on test/functional/feature_bind_port_externalip.py:
```diff
@@ -0,0 +1,76 @@
+            help="Run the test, assuming {} is configured on the machine".format(addr),
+            default=False)
+
+    def setup_nodes(self):
```
Does this need to be overwritten? You can use `skip_test_if_missing_module`

vasild replied:
`raise SkipTest()` is not the main reason to override `setup_nodes()`. It is necessary to specify the command line arguments for each bitcoind node. The current test does that by `self.add_nodes(..., extra_args=my_variable_here)` in `setup_nodes()`. Other tests do that either in `setup_nodes()` or `setup_network()`.

Is there a better way?

MarcoFalke replied:
Set `self.extra_args = ...`?

vasild replied:
Indeed! Changed, thanks for the suggestion!

MarcoFalke commented on test/functional/feature_bind_port_externalip.py:
```diff
@@ -0,0 +1,76 @@
+addr = '1.1.1.1'
+
+# array of tuples [arguments, expected port in localaddresses]
+expected = [
```
uppercase constants

vasild replied:
Changed

MarcoFalke commented on src/init.cpp:
```diff
@@ -1944,11 +1942,15 @@ bool AppInitMain(const util::Ref& context, NodeContext& node, interfaces::BlockA
 
+    // Port to bind to if `-bind=addr` is provided without a `:port` suffix.
+    const uint16_t default_bind_port =
+        static_cast<uint16_t>(gArgs.GetArg("-port", Params().GetDefaultPort()));
```
```suggestion
        static_cast<uint16_t>(args.GetArg("-port", Params().GetDefaultPort()));
```

vasild replied:
Changed

MarcoFalke commented on src/test/net_tests.cpp:
```diff
@@ -677,8 +682,80 @@ BOOST_AUTO_TEST_CASE(ipv4_peer_with_ipv6_addrMe_test)
+    BOOST_CHECK(std::any_of(peer_in.vAddrToSend.begin(), peer_in.vAddrToSend.end(),
+                            [&peer_us](const CAddress& cur) { return cur == peer_us; }));
+
+    gArgs.ForceSetArg("-bind", "");
```
would be good to remove the global here, but that can be done in a follow-up

vasild replied:
Changed to `m_node.args->ForceSetArg("-bind", "");`

MarcoFalke commented on src/test/net_tests.cpp:
```diff
@@ -769,4 +846,62 @@ BOOST_AUTO_TEST_CASE(PoissonNextSend)
+    CDataStream recv_stream{msg.data, SER_NETWORK, PROTOCOL_VERSION};
+
+    // Force CChainState::IsInitialBlockDownload() to return false.
+    const auto nMinimumChainWork_orig = nMinimumChainWork;
+    const auto nMaxTipAge_orig = nMaxTipAge;
+    nMinimumChainWork = 0;
+    nMaxTipAge = GetTime();
```
Can be written shorter

```suggestion
    generatetoaddress(m_node, ADDRESS_BCRT1_UNSPENDABLE);
```

vasild replied:
`CChainState::IsInitialBlockDownload()` still returns `true` :/

(it was necessary to `s/TestingSetup/RegTestingSetup/` in `BOOST_FIXTURE_TEST_SUITE()` in order to get `generatetoaddress()` to accept `ADDRESS_BCRT1_UNSPENDABLE`)

MarcoFalke replied:
Oh, sorry my bad. The mined block will have the timestamp of genesis (plus one), so the time is still too far back.

An alternative would be to use mocktime, also ugly, but less LOC.

```diff
diff --git a/src/test/net_tests.cpp b/src/test/net_tests.cpp
index 9519f3fd0c..7e77b758b2 100644
--- a/src/test/net_tests.cpp
+++ b/src/test/net_tests.cpp
@@ -88,7 +88,7 @@ static CDataStream AddrmanToStream(CAddrManSerializationMock& _addrman)
     return CDataStream(vchData, SER_DISK, CLIENT_VERSION);
 }
 
-BOOST_FIXTURE_TEST_SUITE(net_tests, TestingSetup)
+BOOST_FIXTURE_TEST_SUITE(net_tests, RegTestingSetup)
 
 BOOST_AUTO_TEST_CASE(cnode_listen_port)
 {
@@ -879,11 +879,10 @@ BOOST_AUTO_TEST_CASE(initial_advertise_from_version_message)
 
     CDataStream recv_stream{msg.data, SER_NETWORK, PROTOCOL_VERSION};
 
-    // Force CChainState::IsInitialBlockDownload() to return false.
-    const auto nMinimumChainWork_orig = nMinimumChainWork;
-    const auto nMaxTipAge_orig = nMaxTipAge;
-    nMinimumChainWork = 0;
-    nMaxTipAge = GetTime();
+    // Force CChainState::IsInitialBlockDownload() to return false, otherwise PushAddress isn't called
+    SetMockTime(1);
+    BOOST_REQUIRE(!m_node.chainman->ActiveChainstate().IsInitialBlockDownload());
+    SetMockTime(0);
 
     m_node.peerman->InitializeNode(&peer);
 
@@ -898,9 +897,6 @@ BOOST_AUTO_TEST_CASE(initial_advertise_from_version_message)
     BOOST_CHECK(std::any_of(peer.vAddrToSend.begin(), peer.vAddrToSend.end(),
                             [&expected](const CAddress& cur) { return cur == expected; }));
 
-    nMinimumChainWork = nMinimumChainWork_orig;
-    nMaxTipAge = nMaxTipAge_orig;
-
     gArgs.ForceSetArg("-bind", "");
 }

vasild replied:
Changed, less LOC. Thanks!

maflcko reviewed (commented):
left some style test nits

vasild commented:
Addressed review suggestions and added a test to cover case `4.` from https://github.com/bitcoin/bitcoin/issues/20184.

MarcoFalke commented on test/functional/feature_bind_port_discover.py:
```diff
@@ -0,0 +1,75 @@
+
+    def setup_nodes(self):
+        if not self.options.ihave1111and2222:
+            raise SkipTest(
```
same feedback here ;)

vasild replied:
Changed

vasild commented:
Addressed review suggestions.

vasild commented:
Fixed linter error `feature_bind_port_discover.py:25:18: E703 statement ends with a semicolon` :facepalm:

vasild commented on test/functional/feature_bind_port_externalip.py:
```diff
@@ -0,0 +1,72 @@
+
+# array of tuples [arguments, expected port in localaddresses]
+EXPECTED = [
+    [['-externalip=2.2.2.2',       '-port=20001'                               ], 20001],
```
The ports used better not be hardcoded but derived from `p2p_port()` in order not to clash with other tests that may be running on the system at the same time.

vasild replied:
Changed to start in a fresh range from 30000 as that is the easiest (maximum that can be returned by `p2p_port()` or `rpc_port()` is `PORT_MIN + 2 * PORT_RANGE` which his `21000` :eyes:).

vasild commented:
Rebased and changed port numbers used in the functional tests.

MarcoFalke commented on src/test/net_tests.cpp:
```diff
@@ -926,4 +1002,59 @@ BOOST_AUTO_TEST_CASE(node_eviction_test)
+
+    // Force CChainState::IsInitialBlockDownload() to return false.
+    // Otherwise PushAddress() isn't called by PeerManager::ProcessMessage().
+    SetMockTime(1);
+    BOOST_REQUIRE(!m_node.chainman->ActiveChainstate().IsInitialBlockDownload());
+    SetMockTime(0);
```
nit: Might be possible to write

```suggestion
TestChainState& chainstate = *(TestChainState*)&m_node.chainman->ActiveChainstate();
chainstate.JumpOutOfIbd();
```

vasild replied:
Done. This cast to a class which it actually is not looks quite hackish.

vasild commented:
3e55fbfe3...5647ca32e: rebased due to conflicts + address a [suggestion](https://github.com/bitcoin/bitcoin/pull/20196#discussion_r556430946).

vasild commented:
5647ca32e...58755690a: rebase, what's up with appveyor?

hebasto commented:
@vasild 
> what's up with appveyor?

What do you mean?

vasild commented:
@hebasto https://ci.appveyor.com/project/DrahtBot/bitcoin/builds/37305037

MarcoFalke commented on src/test/net_tests.cpp:
```diff
@@ -929,4 +1004,57 @@ BOOST_AUTO_TEST_CASE(node_eviction_test)
+    std::chrono::microseconds time_received_dummy{0};
+
+    m_node.peerman->ProcessMessage(peer, NetMsgType::VERSION, recv_stream, time_received_dummy,
+                                   interrupt_dummy);
```
I presume this line will make appveyor etc fail if the test process isn't restarted in between test cases.

MarcoFalke replied:
See

```
Test cases order is shuffled using seed: 193951562
```

https://cirrus-ci.com/task/5953114714931200?command=ci#L3792

Which should make it reproducible

vasild replied:
Thanks, that helped a lot nailing the issue down. Fixed now.

vasild committed:
timedata: make it possible to reset the state

Add a new function `TestOnlyResetTimeData()` which would reset the
internal state used by `GetTimeOffset()`, `GetAdjustedTime()` and
`AddTimeData()`.

This is needed so that unit tests that call `AddTimeData()` can restore
the state in order not to confuse other tests that rely on it.

Currently `timedata_tests/addtimedata` is the only test that modifies
the state (via `AddTimeData()`) and also the only test that relies on
that state.

vasild committed:
timedata: rename variables to match the coding style

Rename the local variables in `src/timedata.cpp`:
`setKnown` -> `g_sources`
`vTimeOffsets` -> `g_time_offsets`
`fDone` -> `g_warning_emitted`

vasild commented:
58755690a...d8c6dc0b6: undo the effects of `AddTimeData()` called indirectly by the newly added test `net_tests/initial_advertise_from_version_message` so that `timedata_tests/addtimedata` is not confused. Thanks, @MarcoFalke!

MarcoFalke commented on src/test/timedata_tests.cpp:
```diff
@@ -96,9 +96,10 @@ BOOST_AUTO_TEST_CASE(addtimedata)
     BOOST_CHECK_EQUAL(GetTimeOffset(), 0);
 
-    // We want this test to end with nTimeOffset==0, otherwise subsequent tests of the suite will fail.
+    ResetTimeData();
```
in 9d8bb13b1dfc5103e2678ba803ecd3220dce052d: Shouldn't the reset be before the check equal?

vasild replied:
I don't think so. The `BOOST_CHECK_EQUAL(GetTimeOffset(), 0)` check is to ensure that the net result from all the previous calls is 0. This is still true and better keep that check to ensure it remains so. Surely the check will also pass after `ResetTimeData()` not because the net result is 0, but because the reset obliterated the entire state.

MarcoFalke commented on src/timedata.cpp:
```diff
@@ -106,3 +107,12 @@ void AddTimeData(const CNetAddr& ip, int64_t nOffsetSample)
     }
 }
+
+void ResetTimeData()
```
Could rename this to `TestOnlyResetTimeData`?

vasild replied:
Renamed. I would have preferred to surround it in `#ifdef UNIT_TESTS_ARE_ENABLED` but couldn't find such a macro.

vasild commented:
d8c6dc0b6...70cd1eaef: address [suggestion](https://github.com/bitcoin/bitcoin/pull/20196#discussion_r559688334)

vasild commented:
70cd1eaef...698590c57: rebase due to conflicts

vasild commented:
`698590c57...400cbd480`: rebase due to conflicts

vasild commented:
`400cbd480...b6fb4b928`: rebase due to conflicts

jonatack commented:
Concept ACK

vasild commented:
`b6fb4b928...0c24971f9`: rebase due to conflicts

jonatack reviewed (commented):
Approach ACK after going through each commit and running the new tests. Some suggestions, feel free to pick and choose. There are a couple of things I'd like to verify in the net code when it's not a late hour but globally looks good. I did see a couple of failures in the new functional tests on first run, but they seemed stable after that. (Edit: as these new functional tests will probably only be run manually and not by the CI, it may not be an issue anyway. )

<details><summary>one test failure</summary><p>

```
$ test/functional/feature_bind_port_discover.py --ihave1111and2222
2021-03-17T21:48:41.949000Z TestFramework (INFO): Initializing test directory /tmp/bitcoin_func_test_vh3luuec
2021-03-17T21:48:42.464000Z TestFramework (ERROR): Assertion failed
Traceback (most recent call last):
  File "/home/jon/projects/bitcoin/bitcoin/test/functional/test_framework/test_framework.py", line 523, in start_nodes
    node.wait_for_rpc_connection()
  File "/home/jon/projects/bitcoin/bitcoin/test/functional/test_framework/test_node.py", line 224, in wait_for_rpc_connection
    raise FailedToStartError(self._node_msg(
test_framework.test_node.FailedToStartError: [node 1] bitcoind exited with status 1 during initialization

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/home/jon/projects/bitcoin/bitcoin/test/functional/test_framework/test_framework.py", line 127, in main
    self.setup()
  File "/home/jon/projects/bitcoin/bitcoin/test/functional/test_framework/test_framework.py", line 278, in setup
    self.setup_network()
  File "/home/jon/projects/bitcoin/bitcoin/test/functional/test_framework/test_framework.py", line 372, in setup_network
    self.setup_nodes()
  File "/home/jon/projects/bitcoin/bitcoin/test/functional/test_framework/test_framework.py", line 397, in setup_nodes
    self.start_nodes()
  File "/home/jon/projects/bitcoin/bitcoin/test/functional/test_framework/test_framework.py", line 526, in start_nodes
    self.stop_nodes()
  File "/home/jon/projects/bitcoin/bitcoin/test/functional/test_framework/test_framework.py", line 541, in stop_nodes
    node.stop_node(wait=wait, wait_until_stopped=False)
  File "/home/jon/projects/bitcoin/bitcoin/test/functional/test_framework/test_node.py", line 323, in stop_node
    self.stop(wait=wait)
  File "/home/jon/projects/bitcoin/bitcoin/test/functional/test_framework/test_node.py", line 183, in __getattr__
    assert self.rpc_connected and self.rpc is not None, self._node_msg("Error: no RPC connection")
AssertionError: [node 1] Error: no RPC connection
2021-03-17T21:48:42.515000Z TestFramework (INFO): Stopping nodes
Traceback (most recent call last):
  File "/home/jon/projects/bitcoin/bitcoin/test/functional/feature_bind_port_discover.py", line 74, in <module>
    BindPortDiscoverTest().main()
  File "/home/jon/projects/bitcoin/bitcoin/test/functional/test_framework/test_framework.py", line 151, in main
    exit_code = self.shutdown()
  File "/home/jon/projects/bitcoin/bitcoin/test/functional/test_framework/test_framework.py", line 294, in shutdown
    self.stop_nodes()
  File "/home/jon/projects/bitcoin/bitcoin/test/functional/test_framework/test_framework.py", line 541, in stop_nodes
    node.stop_node(wait=wait, wait_until_stopped=False)
  File "/home/jon/projects/bitcoin/bitcoin/test/functional/test_framework/test_node.py", line 323, in stop_node
    self.stop(wait=wait)
  File "/home/jon/projects/bitcoin/bitcoin/test/functional/test_framework/coverage.py", line 47, in __call__
    return_val = self.auth_service_proxy_instance.__call__(*args, **kwargs)
  File "/home/jon/projects/bitcoin/bitcoin/test/functional/test_framework/authproxy.py", line 144, in __call__
    response, status = self._request('POST', self.__url.path, postdata.encode('utf-8'))
  File "/home/jon/projects/bitcoin/bitcoin/test/functional/test_framework/authproxy.py", line 107, in _request
    self.__conn.request(method, path, postdata, headers)
  File "/usr/lib/python3.9/http/client.py", line 1255, in request
    self._send_request(method, url, body, headers, encode_chunked)
  File "/usr/lib/python3.9/http/client.py", line 1301, in _send_request
    self.endheaders(body, encode_chunked=encode_chunked)
  File "/usr/lib/python3.9/http/client.py", line 1250, in endheaders
    self._send_output(message_body, encode_chunked=encode_chunked)
  File "/usr/lib/python3.9/http/client.py", line 1010, in _send_output
    self.send(msg)
  File "/usr/lib/python3.9/http/client.py", line 950, in send
    self.connect()
  File "/usr/lib/python3.9/http/client.py", line 921, in connect
    self.sock = self._create_connection(
  File "/usr/lib/python3.9/socket.py", line 843, in create_connection
    raise err
  File "/usr/lib/python3.9/socket.py", line 831, in create_connection
    sock.connect(sa)
ConnectionRefusedError: [Errno 111] Connection refused
[node 1] Cleaning up leftover process
[node 0] Cleaning up leftover process
```
</p></details>

jonatack commented on src/net.cpp:
```diff
@@ -113,6 +113,32 @@ void CConnman::AddAddrFetch(const std::string& strDest)
+        NetWhitebindPermissions whitebind;
+        bilingual_str error;
+        if (NetWhitebindPermissions::TryParse(whitebind_arg, whitebind, error)) {
+            if (!(whitebind.m_flags & PF_NOBAN)) {
```
39b4a0c maybe
```suggestion
            if (!NetPermissions::HasFlag(whitebind.m_flags, NetPermissionFlags::PF_NOBAN)) {
```

jonatack replied:
Created a specific `HasAnyFlag` method for this in #21506 at https://github.com/bitcoin/bitcoin/pull/21506/commits/26ac2efcf0acf9bf635f84a9ed986b3d2ddb6042#diff-7016468d59f51adee65131da31d7bebff9818b4e23bb0fac7ab69431b6be4c81R55

jonatack commented on src/test/net_tests.cpp:
```diff
@@ -725,8 +731,79 @@ BOOST_AUTO_TEST_CASE(ipv4_peer_with_ipv6_addrMe_test)
 
+BOOST_AUTO_TEST_CASE(get_local_addr_for_peer_port)
+{
+    // Test that GetLocalAddrForPeer() properly selects the address for self-advertise:
```
39b4a0c309c5f63bdf0bdbde0d76ee3edce3ac19 "for self-advertisement" or "to self-advertise"

vasild replied:
Done.

jonatack reviewed (commented):
Looked at this again today, verified a few additional aspects, and it is really good work.

ACK 0c24971f97d5e6b03aef4911a1f772b6ccfcd6b7 modulo any of the comments you'd like to take

vasild commented:
`0c24971f9...4977c0fdb`: rebase due to conflicts

vasild commented on src/timedata.cpp:
```diff
@@ -70,26 +72,25 @@ void AddTimeData(const CNetAddr& ip, int64_t nOffsetSample)
-    if (vTimeOffsets.size() >= 5 && vTimeOffsets.size() % 2 == 1) {
-        int64_t nMedian = vTimeOffsets.median();
-        std::vector<int64_t> vSorted = vTimeOffsets.sorted();
+    if (g_time_offsets.size() >= 5 && g_time_offsets.size() % 2 == 1) {
```
> as it is done only partially

The commit does not claim to rename all variables, just the 3 ones mentioned in the commit message.

> the header file isn't updated

What do you think needs to be updated in the header file (I assume `timedata.h`)? The variables that were renamed are `static` in `timedata.cpp`.

> 2 occurrences of `vTimeOffsets` in the comments were not updated

Updated the comments, thanks!

vasild commented on test/functional/feature_bind_port_externalip.py:
```diff
@@ -0,0 +1,75 @@
+    [['-externalip=2.2.2.2',       '-port=30016', '-bind={}:30017'.format(ADDR),
+                                             '-whitebind={}:30018'.format(ADDR)], 30017],
+    [['-externalip=2.2.2.2',       '-port=30019',
+                                             '-whitebind={}:30020'.format(ADDR)], 30020],
```
Done.

vasild commented on src/test/net_tests.cpp:
```diff
@@ -945,4 +1022,63 @@ BOOST_AUTO_TEST_CASE(node_eviction_test)
+               CAddress{},
+               std::string{},
+               ConnectionType::OUTBOUND_FULL_RELAY,
+               false};
```
Done.

vasild commented on src/test/net_tests.cpp:
```diff
@@ -945,4 +1022,63 @@ BOOST_AUTO_TEST_CASE(node_eviction_test)
+
+    // Force CChainState::IsInitialBlockDownload() to return false.
+    // Otherwise PushAddress() isn't called by PeerManager::ProcessMessage().
+    TestChainState& chainstate = *(TestChainState*)&m_node.chainman->ActiveChainstate();
```
Done.

vasild commented on src/net_processing.cpp:
```diff
@@ -2465,6 +2465,10 @@ void PeerManagerImpl::ProcessMessage(CNode& pfrom, const std::string& msg_type,
                 } else if (IsPeerAddrLocalGood(&pfrom)) {
+                    // Override just the address with whatever the peer sees us as.
+                    // Leave the port in addr as it was returned by GetLocalAddress()
+                    // above because this is outbound connection and the peer cannot
```
Done.

vasild commented on src/net.cpp:
```diff
@@ -207,7 +233,17 @@ std::optional<CAddress> GetLocalAddrForPeer(CNode *pnode)
+            // For outbound connections assume just the address as seen from
+            // the peer and leave the port in `addrLocal` as returned by
+            // `GetLocalAddress()` above. The peer has no way to observe our
+            // listening port when we have initiated the connection.
```
Done.

vasild commented on src/net.h:
```diff
@@ -184,7 +184,15 @@ enum class ConnectionType {
 std::string ConnectionTypeAsString(ConnectionType conn_type);
+
+/**
+ * Lookup IP addresses from all interfaces on the machine and add them to the
```
Done.

vasild commented on test/functional/feature_bind_port_discover.py:
```diff
@@ -0,0 +1,74 @@
+    def skip_test_if_missing_module(self):
+        if not self.options.ihave1111and2222:
+            raise SkipTest(
+                "To run this test make sure that {} and {} (a routable addresses) are assigned "
```
Done.

vasild commented on test/functional/feature_bind_port_discover.py:
```diff
@@ -0,0 +1,74 @@
+# Distributed under the MIT software license, see the accompanying
+# file COPYING or http://www.opensource.org/licenses/mit-license.php.
+"""
+Test that -discover does not add all intefaces' addresses if we listen just on some
```
Done.

vasild commented on test/functional/feature_bind_port_discover.py:
```diff
@@ -0,0 +1,74 @@
+# We need to bind to a routable address for this test to exercise the relevant code
+# and also must have another routable address on another interface which must not
+# be named "lo" or "lo0".
+# To set a routable address on the machine use:
```
Done.

vasild commented on test/functional/feature_bind_port_discover.py:
```diff
@@ -0,0 +1,74 @@
+                "to the interfaces on this machine and rerun with --ihave1111and2222".format(ADDR1, ADDR2))
+
+    def run_test(self):
+        # Check that if no -bind= is given, then all addresses are added to localaddresses.
```
Done.

vasild commented on test/functional/feature_bind_port_discover.py:
```diff
@@ -0,0 +1,74 @@
+        assert(found_addr1)
+        assert(found_addr2)
+
+        # Check that if -bind= is given, just that address is added to localaddresses.
```
Done.

vasild commented:
`4977c0fdb...7d5d2416e`: address review suggestions

jonatack commented on test/functional/feature_bind_port_externalip.py:
```diff
@@ -0,0 +1,75 @@
+                "to one of the interfaces on this machine and rerun with --ihave1111")
+
+    def run_test(self):
+        for i in range(len(EXPECTED)):
```
would be good to add the missing logging suggested in https://github.com/bitcoin/bitcoin/pull/20196#discussion_r596410851
```diff
     def run_test(self):
+        self.log.info("Test the proper port is used for -externalip=")
         for i in range(len(EXPECTED)):
             if EXPECTED[i][1] == 'default_p2p_port':
                 expected_port = p2p_port(i)
@@ -69,7 +70,8 @@ class BindPortExternalIPTest(BitcoinTestFramework):
                     assert_equal(local['port'], expected_port)
                     found = True
                     break
-            assert(found)
+            assert found
+
```
nit, remove brackets after the assert and two lines before the `main` section)

vasild replied:
Done.

jonatack reviewed (commented):
ACK 7d5d2416e70925ee65bca455d88427335dc12008 modulo some nits below and a test that accepts the "wrong" option:
```
$ test/functional/feature_bind_port_discover.py
2021-03-30T08:48:27.978000Z TestFramework (WARNING): Test Skipped: To run this test make sure that 1.1.1.1 and 2.2.2.2 (routable addresses) are assigned to the interfaces on this machine and rerun with --ihave1111and2222
2021-03-30T08:48:28.029000Z TestFramework (INFO): Stopping nodes
2021-03-30T08:48:28.029000Z TestFramework (INFO): Cleaning up /tmp/bitcoin_func_test_gsbqg2e2 on exit
2021-03-30T08:48:28.029000Z TestFramework (INFO): Test skipped

$ test/functional/feature_bind_port_discover.py --ihave1111
2021-03-30T08:46:34.603000Z TestFramework (INFO): Initializing test directory /tmp/bitcoin_func_test_b4bp9ak2
2021-03-30T08:46:34.914000Z TestFramework (INFO): Test that if -bind= is not passed then all addresses are added to localaddresses
2021-03-30T08:46:34.916000Z TestFramework (INFO): Test that if -bind= is passed then only that address is added to localaddresses
2021-03-30T08:46:34.967000Z TestFramework (INFO): Stopping nodes
2021-03-30T08:46:35.124000Z TestFramework (INFO): Cleaning up /tmp/bitcoin_func_test_b4bp9ak2 on exit
2021-03-30T08:46:35.125000Z TestFramework (INFO): Tests successful

$ test/functional/feature_bind_port_discover.py --ihave1111and2222
2021-03-30T08:50:04.718000Z TestFramework (INFO): Initializing test directory /tmp/bitcoin_func_test__iqqpdav
2021-03-30T08:50:05.013000Z TestFramework (INFO): Test that if -bind= is not passed then all addresses are added to localaddresses
2021-03-30T08:50:05.015000Z TestFramework (INFO): Test that if -bind= is passed then only that address is added to localaddresses
2021-03-30T08:50:05.070000Z TestFramework (INFO): Stopping nodes
2021-03-30T08:50:05.245000Z TestFramework (INFO): Cleaning up /tmp/bitcoin_func_test__iqqpdav on exit
2021-03-30T08:50:05.245000Z TestFramework (INFO): Tests successful
```

vasild commented:
> ...a test that accepts the "wrong" option

That must be some "feature" in the testing framework - it also accepts `--ihave11`.

vasild commented on src/test/net_tests.cpp:
```diff
@@ -726,8 +732,79 @@ BOOST_AUTO_TEST_CASE(ipv4_peer_with_ipv6_addrMe_test)
 
+BOOST_AUTO_TEST_CASE(get_local_addr_for_peer_port)
+{
+    // Test that GetLocalAddrForPeer() properly selects the address to self-advertis:
```
Fixed.

vasild commented on test/functional/feature_bind_port_discover.py:
```diff
@@ -0,0 +1,78 @@
+# Distributed under the MIT software license, see the accompanying
+# file COPYING or http://www.opensource.org/licenses/mit-license.php.
+"""
+Test that -discover does not add all interfaces' addresses if we listen just on only some of them
```
Changed.

vasild commented:
`7d5d2416e...3501180d3`: rebase due to conflicts

vasild commented:
`3501180d3...33b0a79df`: address suggestions

jonatack reviewed (commented):
Code review ACK 33b0a79df62d99baaf1e37f13a9c52b7bbe3955e per `git range-diff 4ad83a9 7d5d241 33b0a79` to compare with my last review, modulo one comment below

vasild commented on src/net.cpp:
```diff
@@ -117,6 +117,32 @@ void CConnman::AddAddrFetch(const std::string& strDest)
+        NetWhitebindPermissions whitebind;
+        bilingual_str error;
+        if (NetWhitebindPermissions::TryParse(whitebind_arg, whitebind, error)) {
+            if (!(whitebind.m_flags & PF_NOBAN)) {
```
Good catch!

Previously, if `-whitebind=download@1.1.1.1:8765` was specified we would not have returned `8765` from `GetListenPort()` but we should have.

jonatack replied:
Yes, I've been working on this issue with #21644 that fixes the same issue and #21506 that would eliminate this form of `NetPermissionFlags` bug.

Can you please use the `NetPermissionFlags::` scope as suggested in https://github.com/bitcoin/bitcoin/pull/20196#discussion_r610770148? This is already the case in other places and it would avoid an unnecessary future commit (or rebase of #21506).

vasild commented:
`33b0a79df...a5590dcf9`: use `NetPermissions::HasFlag()` to check if `PF_NOBAN` is set, thanks @jonatack!

jonatack commented:
re-ACK a5590dcf9dccd82bb0784349d2a3c91dec263380

Happy to re-ack if you add the `NetPermissionFlags::` scope mentioned in https://github.com/bitcoin/bitcoin/pull/20196#discussion_r612676028.

vasild commented:
`a5590dcf9...7126e12fa`: use `NetPermissionFlags::PF_NOBAN` instead of `PF_NOBAN`

jonatack commented:
re-ACK  7126e12fab1eff95efeaab349a49e4f1f714ed4f  per `git diff a5590dc 7126e12`

jonatack commented:
Bump. This could use a review by someone!

vasild commented:
`7126e12fab...9c15f204e5`: conflictless rebase to get green CI

vasild commented:
`9c15f204e5...2519696300`: fix silent merge conflict

jonatack commented on src/net.cpp:
```diff
@@ -118,6 +118,32 @@ void CConnman::AddAddrFetch(const std::string& strDest)
+        }
+    }
+
+    // Otherwise, if -whitebind= without "noban" is provided, use that
```
nit, while updating here could be consistent with 7075f604e8d0b21
 ```suggestion
    // Otherwise, if -whitebind= without NetPermissionFlags::NoBan permission is provided, use that
```

vasild replied:
Done.

jonatack reviewed (commented):
ACK 2519696300a356ba2876d37bb01ba620ad46bc0c per `git range-diff d4c409cf0 7126e12 2519696`

happy to re-ACK if you update per the comment

jonatack commented:
Thanks!

Code review re-ACK  b962cbc22cd35a6f7c5b42a0f3a113c109b4cee2 per `git range-diff d4c409c 2519696 b962cbc`

vasild commented:
`b962cbc22c...4ea0a60f75`: rebase due to conflicts

jonatack commented:
Build:
```
test/net_tests.cpp:964:34: error: no member named 'vAddrToSend' in 'CNode'
    BOOST_CHECK(std::any_of(peer.vAddrToSend.begin(), peer.vAddrToSend.end(),
                            ~~~~ ^

test/net_tests.cpp:964:60: error: no member named 'vAddrToSend' in 'CNode'
    BOOST_CHECK(std::any_of(peer.vAddrToSend.begin(), peer.vAddrToSend.end(),
                                                      ~~~~ ^
2 errors generated.
```

vasild commented:
Yes :sob: after recent changes `CNode::vAddrToSend` has been moved to `Peer::m_addrs_to_send` (`Peer` is defined only in `net_processing.cpp`, not exposed outside). So this is now nearly untestable. I am considering how to resolve this. Would hate to drop the test.

vasild committed:
net: make CaptureMessage() mockable

Rename `CaptureMessage()` to `CaptureMessageToFile()` and introduce a
`std::function` variable called `CaptureMessage` whose value can be
changed by unit tests, should they need to inspect message contents.

vasild commented:
`4ea0a60f75...6813052724`: rewrite the test `net_tests/initial_advertise_from_version_message` so that it works with latest `master`

vasild commented:
`6813052724...2aee1b92c5`: fix `net_tests/get_local_addr_for_peer_port`: compare `CService` instead of `CAddress` - "time" and "services" (part of `CAddress`) are irrelevant and should not be compared (they differ).

jonatack reviewed (commented):
Almost re-ACK WIP review, quite a few changes and context to swap back into memory since my last review a few months ago. Some checkpoint comments and notes per commit follow, feel free to ignore.

jonatack commented on src/init.cpp:
```diff
@@ -1745,6 +1747,12 @@ bool AppInitMain(NodeContext& node, interfaces::BlockAndHeaderTipInfo* tip_info)
+    if (connOptions.bind_on_any) {
+        // Only add all IP addresses of the machine if we would be listening on
+        // any address - 0.0.0.0 (IPv4) and :: (IPv6).
+        Discover();
```
2aee1b92 verified that without the change in this commit, the accompanying functional test fails at the expected assertion
```
File "~/bitcoin/test/functional/feature_bind_port_discover.py", line 74, in run_test
    assert local['address'] != ADDR2
AssertionError
```

vasild committed:
net: pass Span by value to CaptureMessage()

Span is lightweight and need not be passed by const reference.

vasild commented:
`2aee1b92c5...ad6c1fbe8f`: address suggestions

vasild commented on src/net.cpp:
```diff
@@ -3102,3 +3141,9 @@ void CaptureMessage(const CAddress& addr, const std::string& msg_type, const Spa
+
+std::function<void(const CAddress& addr,
+                   const std::string& msg_type,
+                   const Span<const unsigned char>& data,
```
Done.

vasild commented on src/init.cpp:
```diff
@@ -1696,11 +1694,15 @@ bool AppInitMain(NodeContext& node, interfaces::BlockAndHeaderTipInfo* tip_info)
         const size_t index = bind_arg.rfind('=');
         if (index == std::string::npos) {
-            if (Lookup(bind_arg, bind_addr, GetListenPort(), false)) {
+            if (Lookup(bind_arg, bind_addr, default_bind_port, false)) {
```
Done.

vasild commented on src/timedata.cpp:
```diff
@@ -106,3 +107,12 @@ void AddTimeData(const CNetAddr& ip, int64_t nOffsetSample)
+    LOCK(g_timeoffset_mutex);
+    nTimeOffset = 0;
+    g_sources.clear();
+    g_time_offsets = CMedianFilter<int64_t>(BITCOIN_TIMEDATA_MAX_SAMPLES, 0);
```
Done.

vasild commented on src/net.cpp:
```diff
@@ -120,6 +120,32 @@ void CConnman::AddAddrFetch(const std::string& strDest)
+                return bind_addr.GetPort();
+            }
+        }
+    }
```
Done.

vasild commented:
`ad6c1fbe8f...e4474f66d9`: change forgotten argument to pass `Span` by value

jonatack commented:
ACK e4474f66d91228b8ddcc8ce4fb5ebdb447ecede5 rebased and debug-built to current master to check for silent merge conflicts, reviewed code, removed the code behind each behavior change and verified the relevant test failed where expected, and passed with the code re-inserted

Possible additional to-do: manually reproduce and verify the issues outlined in https://github.com/bitcoin/bitcoin/issues/20184.

This PR fixes a number of issues--nice work!

vasild commented:
`e4474f66d9...812deb7470`: rebase due to conflicts

sipa commented:
utACK 812deb7470d1d60ee6756b2cbec72bb32999312e. Agree with the concept and code changes. I've only superficially reviewed the new test.

jonatack commented:
Code review re-ACK 812deb7470d1d60ee6756b2cbec72bb32999312e per `git range-diff f8b837 e4474f6 812deb7` rebase-only since my last review

vasild commented:
`812deb7470...cb86c1fd20`: rebase due to conflicts

Invalidates ACKs from @sipa, @jonatack.

jonatack reviewed (approved):
re-ACK cb86c1fd20715c9a7e95dea1fd2ce6894d7e9ad5 per `git range-diff 571bb94 812deb7 cb86c1f` :coffee:

theStack commented:
Concept ACK

jonatack commented:
Tagged for v23, needs rebase.

vasild commented:
`cb86c1fd20...38f49fc11d`: rebase due to conflicts (thanks for the poke!)

Invalidates ACK from @jonatack 

Previously invalidated ACK from @sipa

jonatack commented on src/init.cpp:
```diff
@@ -1698,11 +1696,15 @@ bool AppInitMain(NodeContext& node, interfaces::BlockAndHeaderTipInfo* tip_info)
         const size_t index = bind_arg.rfind('=');
         if (index == std::string::npos) {
-            if (Lookup(bind_arg, bind_addr, GetListenPort(), false)) {
+            if (Lookup(bind_arg, bind_addr, default_bind_port, /* allow_lookup */ false)) {
```
a6d15f47 as you updated the named args in the net tests, if you retouch
```suggestion
            if (Lookup(bind_arg, bind_addr, default_bind_port, /*allow_lookup=*/false)) {
```

vasild replied:
Done.

jonatack reviewed (commented):
re-ACK 38f49fc11d1d184a6c804d46340e4f91f47901df

vasild commented:
`38f49fc11d...35ec977b06`: rebase and use `/*fAllowLookup=*/`

jonatack commented:
re-ACK 35ec977b067d05c318b68b4eafbf708870ba9d3f

jonatack commented:
This has been open for a year and half, has seen a fair amount of review, has ACKs by myself and @sipa, and is tagged for 23.0, if anyone else would like to have a look!

sipa commented:
utACK 7d64ea4a01920bb55bc6de0de6766712ec792a11. I didn't review the tests in detail.

jonatack reviewed (commented):
re-ACK 7d64ea4a01920bb55bc6de0de6766712ec792a11 per `git range-diff 08bcfa27 35ec977 7d64ea4`, changes are rebase-only, light re-review, re-ran the new tests locally

prayank23 commented:
Concept ACK
